pyrometheus/codegen/python.py

```diff
@@ -321,6 +327,33 @@ def get_mixture_internal_energy_mass(self, temperature, mass_fractions):
         emix = self.get_mass_average_property(mass_fractions, e0_rt)
         return self.gas_constant * temperature * emix
 
+    def get_mixture_entropy_mass(self, pressure, temperature, mass_fractions):
+        mmw = self.get_mix_molecular_weight(mass_fractions)
+        return 1.0/mmw * self.get_mixture_entropy_mole(pressure, temperature, <%
+                                                       %>mass_fractions)
+
+    def get_mole_average_property(self, mass_fractions, spec_property):
+        mmw = self.get_mix_molecular_weight(mass_fractions)
+        mole_fracs = self.get_mole_fractions(mmw, mass_fractions)
+        return sum([mole_fracs[i] * spec_property[i]
+                    for i in range(self.num_species)])
+
+    def get_mixture_enthalpy_mole(self, temperature, mass_fractions):
+        h0_rt = self.get_species_enthalpies_rt(temperature)
+        hmix = self.get_mole_average_property(mass_fractions, h0_rt)
+        return self.gas_constant * temperature * hmix
+
+    def get_mixture_entropy_mole(self, pressure, temperature, mass_fractions):
+        mmw = self.get_mix_molecular_weight(mass_fractions)
+        # necessary to avoid nans in the log function below
+        x = self.usr_np.where(
```

Comment:
I am uneasy about this type of stuff being part of the template. Generally, these numerical tricks can be worked in by overriding individual functions of the generated code.  Particularly, the resulting mole fractions `x` do not correspond to the mass fractions passed as inputs.

Reply:
The issue is that the code will simply crash. There is no log of negative species and this is not something we can put outside of pyrometheus.

Reply:
I'm confused now. Do we keep mole-based properties or not? Or are they now in `test_codegen.py`? Let me take a look, but I don't think they belong there. We can have an auxiliary file that adds them to the generated class (I can do that).